velox/expression/SwitchRewrite.cpp

```diff
@@ -0,0 +1,100 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <set>
+
+#include "velox/expression/ExprConstants.h"
+#include "velox/expression/ExprRewriteRegistry.h"
+#include "velox/expression/ExprUtils.h"
+#include "velox/expression/SwitchRewrite.h"
+
+namespace facebook::velox::expression {
+
+core::TypedExprPtr SwitchRewrite::rewrite(const core::TypedExprPtr& expr) {
+  if (!expr->isCallKind()) {
+    return nullptr;
+  }
+  const auto* callExpr = expr->asUnchecked<core::CallTypedExpr>();
+  // `if` is a special form of `switch` with a single condition.
+  if (!(callExpr->name() == kSwitch) && !(callExpr->name() == kIf)) {
+    return nullptr;
+  }
+
+  const auto& inputs = expr->inputs();
+  const auto numInputs = inputs.size();
+  std::vector<core::TypedExprPtr> optimizedInputs;
+  bool hasOptimizedElseClause = false;
+
+  // Iterate over all `(condition, value)` pairs. Handle `else` clause at end.
+  for (auto i = 0; i < numInputs - 1; i += 2) {
+    const auto& condition = inputs.at(i);
+    const auto& value = inputs.at(i + 1);
+
+    if (condition->isConstantKind()) {
+      const auto* constCondition =
+          condition->asUnchecked<core::ConstantTypedExpr>();
+      if (auto boolCondition = constCondition->toBool()) {
+        if (boolCondition.value()) {
+          // If this condition is true and all conditions before this are false,
+          // simplify `switch` expression to the corresponding `value`.
+          if (optimizedInputs.empty()) {
+            return value;
+          }
+          // If this condition is true and all conditions before this can only
+          // be evaluated at runtime, make this the new `else` clause and stop
+          // checking further conditions.
+          hasOptimizedElseClause = true;
+          optimizedInputs.push_back(value);
+          break;
+        } else {
+          // Skip false conditions.
+          continue;
+        }
+      } else {
+        // Skip NULL conditions.
+        continue;
+      }
+    } else {
+      optimizedInputs.push_back(condition);
+      optimizedInputs.push_back(value);
+    }
+  }
+
+  // Handle default `else` clause if present.
+  if (!hasOptimizedElseClause && numInputs % 2 == 1) {
+    const auto elseValue = inputs.at(numInputs - 1);
+    // If all conditions are false, return value from `else` clause.
+    if (optimizedInputs.empty()) {
+      return elseValue;
+    }
+    optimizedInputs.emplace_back(elseValue);
+  }
+  // Return NULL if all conditions are false and no default `else` clause is
+  // present.
+  if (optimizedInputs.empty()) {
+    return std::make_shared<core::ConstantTypedExpr>(
```

Comment:
nit: let's add a static helper method ConstantTypedExpr::makeNull(type) to reduce boiler plate